### DISCUSSION TITLE
Default Settings for Set Token Light and Vision macro

### DISCUSSION THREAD

**Discussion by Jagusti** [Jagusti/fvtt-wfrp4e-gmtoolkit#193] [Ideas]
The next significant release of GM Toolkit is focused on improvements that can take advantage of Foundry v10's updated **vision, lighting and canvas engine**.  

Among those changes I want to revisit the settings I use in the **Set Token Vision and Light** macro. The current defaults are described in this wiki page: https://github.com/Jagusti/fvtt-wfrp4e-gmtoolkit/wiki/set-token-vision-and-light. 

I'd like to get some input about how well these work, and whether there are some config changes I can make to get a better baseline set that works for more people out the box.  
- I'm unlikely to change distances as most are defined in the rules (but you're welcome to persuade me). 
- I am open to changing any other parameter, such as intensity, colour, animation and other settings. 
- I will be exploring options to integrate Vision and Detection modes. 
- I also want to streamline the lighting options to make a more manageable list. 

In this thread, please do share your feedback and input on: 
- **configurations**, eg, settings you use for a light source and vision/detection mode combo
- **effects**, eg, torch-based lights are generally too bright, Pha's Protection animations is too intense
- **how you use the macro**, eg, in my group we mainly use Torch, Lantern and Davrich Lamp, and never use condition related lighting effects

The default effects settings are just one part of the changes I want to make to the Set Token Vision and Light macro, and that's just one canvas related feature that will be tackled as part of this release.  If you have **other ideas** related to light and vision that I might be able to consider as part of this release, 
- feel free to open up another discussion thread 
- or look out for other issues that will appear on the [upcoming milestone plan](https://github.com/Jagusti/fvtt-wfrp4e-gmtoolkit/milestone/11)

Thanks for reading this far, and for any input you can offer.

**Comment by TheFirst05** [comment]
Hi. We use the set token vision macro a lot, It's the one macro that everyone at the table uses. We mostly uses torches, candles and lanterns. I think I used matches once, but we haven't really used the others.

The best enhancement to the macro itself would be a having 2-3 buttons instead of drop down box. Something where we can click our most common options (usually being no light, torch, lantern). Currently there are so many options in the dropdown box that you need to focus to select the option you want. With some buttons, it's a much simpler interaction. I'm no UI designer though, I don't know what you'd do with all the other options. I don't know if there's like some true/false arguments you can put in the macro so we can selectively add/remove some options without just deleting the lines of code.

As a personal preference, we've decreased the colour intensity and animation Intensity/Speed settings across the board. The intensity was washing out just a bit too much colour, and the animations were a bit too distracting. As a general rule of thumb, I think I put Torch as "`lightColorIntensity = 0.3; animationIntensity =3; animationSpeed = 3;`. I could post every single setting, but honestly most of them were just arbitrarily decreased. Most animations don't go above 4 (especially speed, which is the setting I found most distracting), and the colour intensity is around 0.2-0.5.

One thing we'd love to see is a Detection Mode that truly simulates Night Vision. We currently have a gnome in the party, but they almost never use the Night vision option in the macro, because they aren't carrying the torch. I have no idea if its possible, but making it so that Night Vision characters can see 20 yards further would be shining achievement for lighting in wfrp. Honestly this feature is something I think could belong in the WFRP4e game system, but at the end of the day I don't mind which module I need to download to get this feature 😄.

As an aside, I don't know if you should be able to see completely monochrome with Nightvision. It isn't exactly like Darkvision in D&D, where you see black and white in total darkness.

Sorry if this feedback is a bit of a mess. A lot is personal preference at the end of the day.

**Reply by Jagusti** [reply]
Thanks for this ... no need to apologise, it's not messy feedback, but very helpful. Also aligns with some of the instincts I have for the changes to be made: 

> best enhancement to the macro itself would be a having 2-3 buttons instead of drop down box

I have a few plans for the user interface: 
- overhaul the UI to align with the Damage and Group Test console. That UI pattern is emerging nicely for the Toolkit, and supports 4 x 1-click button options as well as a more extended list and options if you need to do something less usual. 
- explore user settings that allow you to customise what lighting configurations are shown, so even though there's 15 pre-configured light options, you still may only want 6-7 in the dropdown, even with the quick apply buttons.  This would be additional to allowing GMs to adjust the default light configuration settings. 
- Stretch goal that could help improve the UI would be to intelligently default light source selection based on what the token is carrying.  Gets tricky if they have candles, matches and lanterns ... so this may be a thing too far. 
- What should be readily doable is defaulting sight based on vision traits/talents and conditions as per
  - #41 
  - #42

> we've decreased the colour intensity and animation Intensity/Speed settings across the board

Agreed, and this is consistent with what others are reporting.  As well as reducing those settings, I'll be experimenting with different Coloration Techniques. Adaptive Luminance is the default, but seems too gaudy. For sources such as matches or candles, High or Low Absorption may fit better. 

> a Detection Mode that truly simulates Night Vision

Yep, that's where I'd like to be but I should say it's going to be a stretch for me.  One that I'm likely to need more help with grokking how they work inside Foundry. 

> making it so that Night Vision characters can see 20 yards further would be shining achievement

That's what the macro already does, as long is it's not 100% Darkness. It extends the token's sight range by 20 yds per rank of Night Vision, plus whatever the Dim range of their light source is, which may be 0 if they aren't carrying anything. So they'll always get at least 20yds.  Please do raise a bug if that's not happening. 

> monochrome with Nightvision

Yea, that's gonna change ... I got too eager with the new Vision Modes. :)

**Comment by Txus5012** [comment]
Hi. We mostly use the vision macro for torches, storm lanterns and similar. Here are my ideas.

- It's great to have the wide and narrow versions of the storm lantern, but the closed one feels useless. In those cases I just use no light.
- For candles, torches and similar the animation maybe should be now "flame".
- I agree with TheFirst05 that the lights are now too intense, but it's also true that editing the macro is easy.
- I also agree that Night Vision shouldn't be monochrome or blue. It was pretty good before, like basic vision but with more range.
- I've been fantasizing with a detection mode for Second Sight, with an effect showing if a token is suffused with magic. Maybe a new condition? Maybe Detect Invisibility can be a good inspiration?

I'm working in adding (optional) light for the Ablaze condition in my WFRP4e-FX module. For the blinded condition v10 has a nice blinded effect and I opened an issue about how to apply it in the system [here](https://github.com/moo-man/WFRP4e-FoundryVTT/issues/1314).

**Reply by Jagusti** [reply]
Thanks for these @Txus5012 - also really useful. 

> It's great to have the wide and narrow versions of the storm lantern, but the closed one feels useless. In those cases I just use no light.

An option to customise the light source menu will help here.

> For candles, torches and similar the animation maybe should be now "flame".

Good catch 👍🏽 

> I agree with TheFirst05 that the lights are now too intense, but it's also true that editing the macro is easy.
> I also agree that Night Vision shouldn't be monochrome or blue. It was pretty good before, like basic vision but with more range.

Yep - see my answers to TheFirst

> I've been fantasizing with a detection mode for Second Sight, with an effect showing if a token is suffused with magic. Maybe a new condition? Maybe Detect Invisibility can be a good inspiration?

A stretch goal for the next update, but if I crack Detection Modes, I'd like to try this, for sure:
- There's probably other existing properties to check for, such as having Channelling, Corruption, Mutations, rather than adding a new Condition. 
- The useful thing for me with Second Sight is showing a trail - that might be harder (unless the character's movement is tracked ... I don't want to go there!).
- I like the idea of being able to customise the colour of the sight based on the Second Sight owner's wind ... hopefully that's doable too. 

> I'm working in adding (optional) light for the Ablaze condition in my WFRP4e-FX module. 

Saw that got released - nice one 👍🏽 

> For the blinded condition v10 has a nice blinded effect and I opened an issue about how to apply it in the system https://github.com/moo-man/WFRP4e-FoundryVTT/issues/1314.

Cool - have subscribed.

**Reply by Jagusti** [reply]
> It's great to have the wide and narrow versions of the storm lantern, but the closed one feels useless.

Oh yes, the `Storm Lantern (Shuttered)` option ... true, at 0yds it does feel redundant at the moment. I guess where it's handy is for those who want to know that there is a light in play, but it's effectively switched off.  Of course, that info isn't surfaced at the moment ... but there's something to consider here ... 

In the same way you can equip a weapon, maybe it should be possible to separately equip and light / extinguish item-based light sources such as matches, candles, torches and lanterns.  There's no in-game ruling about how something like that would work, but this is an interesting case of where the VTT support a pretty cool playing experience if the system enabled it. 

In this case, a new pseudo-condition for equipped light could work really rather nicely: active being on, passive being equipped but off. 

Fair to say, this would make an interesting information architecture and user interface problem.

**Comment by fremen666** [comment]
Hi, Im fremen (same nickname on discord vtt server) So my comment on vision:
I do not know direct insights of vtt system, but I will try to use my everyday gamedev knowledge to try to sum it up from my perspective:
Rules:
From what i can see there are 3 rules for vision in wfrp 4e:
1. everyone can see as far as they can when there is enough light source
2. everyone who has Night Vision Talent can see specific amount of feet when in "low light conditions" like moon
3. everyone who possess Darkvision can see in specific amount of feet when in total darkness.

For total darkness we can probably agree. From normal life experience, no light at all, you can't even see your own hand in front. You can experience it by entering very dark places like bunkers.

For low light conditions it is trickier because for each person this may mean something else but for something like vtt simulation we can do some assumptions.

What do we have in vtt system from what I understand:
There is Global illumination that can be set on the scene that sets the "default" light level on whole scene. This is when if scene darkness level value is lower than Global Illumination threshold.
Otherwise, darkness level value is just used currently for visual aspect. 
There is "Vision Range" property under Vision tab on the token that currently is used for vision range when GI is on and Darkness level is above GI threshold.

My proposition:
1. we add additional property next to "Vision Range" that would be called Night Vision range or low light range (name up to you). If possible, in UI it would be next to the current Vision Range value but if not than it would be anywhere else settable (up to you)
2. instead of having 1 value "GI threshold" we would have "GI low threshold" and "GI high threshold". "GI high threshold" would be essentially current GI threshold.
3. We use scene Darkness Level value as "global light" value instead of just visual. 

Now if DL (darkness level) is between "GI low threshold" and "GI high threshold" we use new Night Vision value for range
if DL is above "GI high threshold" we use "Vision Range" as currently to determine vision range
if DL is below "GI low threshold" we assume token has full unlimited range of vision as the scene is lit properly.

This logic of vision should be applied to the token when selected (plus any light on token) and for lights it should apply their vision range + token vision after determining it from above logic.

I hope it is clear. If not feel free to ask me questions directly on discord. Im not sure if i will be aware of them here but ill try :)

**Comment by TheFirst05** [comment]
Hi Jagusti, I don't know if you've had any time to work on Lighting and Vision, but I've got something you might (or might not!) find useful. 

Below is the current implementation of "Low-light vision" in the Pathfinder 1e system. In that system, tokens with Low-light vision can see twice as far with a source of light. In the video below, I have a light source and two tokens, the top one with Basic Vision out to 2 yards, and the bottom token with 2 yards basic vision AND the Low-Light vision trait. The Game system has options in both the light's setting and the token's setting to disable this functionality.
![Low Light Vision in Pathfinder](https://user-images.githubusercontent.com/96636889/212477913-3eadac14-9a89-4fdb-a9b5-c76d68c60b4c.gif)

![image](https://user-images.githubusercontent.com/96636889/212479372-caa2c2d6-b006-4fae-ae36-bf94319b7568.png)


The Night vision Talent has 2 main effects (I will ignore the effect of multiple levels in the talent for now):

- you can see 20 yards as long as you have a very faint source of light (starlight, etc...)
- you can extend the effective illumination of any light source by 20 yards.

While the first effect might be a bit vague on what that actually means (how much light is enough?), the second effect is (in my opinion), very cut and dry to implement: if a light source exists on the map, you can increase its effective illumination by 20 yards (so a lantern is 40 yards bright). This second effect would perfectly compliment (or even replace?) the current implementation of Night Vision in the Set Token Light and Vision Macro, especially for those cave /dungeon maps where only one person is carrying a source of light, or if the DM has placed candles in certain areas.

I don't know if its possible to reverse engineer the Low-light Vision code from Pathfinder into the WFRP4e system, but I thought I should maybe mention it, if you were not aware of it already. 

In any case, thanks a lot for the work you've been doing on this module.

**Reply by Jagusti** [reply]
Thanks for this ... very interesting, and perhaps the closest I've seen to the most likely interpretation of night vision rules ... you are also correct to say that I've not really had time to make the progress I hoped in this area. I'll rememeber this for when I get back to it though.

**Comment by Jagusti** [comment]
Plenty more to do, but finally got a chance to do the first round of changes focusing on making the effects less visually distracting. Now available in [v6.0.3](https://github.com/Jagusti/fvtt-wfrp4e-gmtoolkit/discussions/220). Thanks for input so far ... feedback welcome on the updates so far.

**Reply by TheFirst05** [reply]
Thanks for the changes! Agreed it's a lot less distracting. A lot easier on the eyes.

I really like how the Ablaze/Soulfire lights look. Great use of the torch animation.

At our table, personal opinion is we prefer using the Flickering Lights animation for matches, candles and torches. Otherwise it feels like our character's hair/faces are on fire. Regardless, it's an easy change to make on our end. 

I'm really grateful that we have a well documented light macro that is easy to follow, with all the baseline values set out and the different modifications for each light. Makes it really simple to tweak depending on GM/player/table preference.

A question: the "Normal Vision Range" in the Module's settings, is it possible to set it so we can input a value between integers? We're experimenting with having vision in darkness expand to 0.5 yards (or similar) instead of 2 yards. That's mainly because we use 1 yard square maps instead of the default 2 yards.

**Reply by Jagusti** [reply]
Appreciate the feedback ... thanks @TheFirst05 ... glad to hear that it's improved. And that the it's been worthwhile for someone other than me to maintaining good docs :) 

- I'll take another look at torch v flickering for flame-based light sources. There may be scope to tone it down.

- A bit like when I updated Advantage to pave the way for group advantage automation, I ended up pushing a lot of the functionality outside of the macro and into functions behind the scenes. I'm conscious that with the changes that I expect to make for Token Vision and Light, some of this may make it harder to tweak than it currently is, as it won't be going directly via the macro body. I'll keep in mind the value of making that easy, and will see what I can keep in place that doesn't compromise that too much. 

- Looks like all my non-range number based settings are integer-protected via the UI, which is useful most of the time, but not for your case. I'm struggling to fathom why I can't override that (I'll need to tap into the Foundry hive mind for more help), but if you enter the following into the console you should be able to adjust it with fractions: 
`await game.settings.set("wfrp4e-gm-toolkit", "rangeNormalSight", Number(0.5))`

**Reply by Jagusti** [reply]
> is it possible to set it so we can input a value between integers? We're experimenting with having vision in darkness expand to 0.5 yards (or similar) instead of 2 yards. 

Added an enhancement for this that will go out in the next release: #223

**Reply by TheFirst05** [reply]
> I'll take another look at torch v flickering for flame-based light sources. There may be scope to tone it down.

At the end of the day it's no big deal. Other tables might prefer it. My table has some pretty strong opinions sometimes. The last thing I want to do is shower you with small tweaks to suit our specific table's desires. As I said, the macro is easy to personalise (especially the lights part, which is the most subjective to taste).

> Added an enhancement for this that will go out in the next release: https://github.com/Jagusti/fvtt-wfrp4e-gmtoolkit/issues/223

Thanks a lot for this! I've eventually worked out where the macro was looking for sightRange value, which means I could have also modified the macro to display 0.5 instead of `Number(game.settings.get("wfrp4e-gm-toolkit", "rangeNormalSight"))`, but this enhancement will give other tables the option to use less conventional ranges.
